sdk/src/wallet/utxo.ts

```diff
@@ -179,3 +214,165 @@ export function getInputFromUtxoAndTx(
+ * const fromAddress = 'bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq';
+ * const publicKey = '02d4...`; // only for P2SH
+ * const opReturnData = 'Hello, World!'; // optional
+ * const feeRate = 1; // optional
+ *
+ * const fee = await estimateTxFee(fromAddress, undefined, publicKey, opReturnData, feeRate);
+ * console.log(fee);
+ * ```
+ *
+ * @dev Wtih no amount set, we estimate the fee for all UTXOs by trying to spend all inputs using strategy 'all'. If an amount is set, we use the 'default
+ * strategy to select the UTXOs.
+ */
+export async function estimateTxFee(
+    fromAddress: string,
+    amount?: number,
+    publicKey?: string,
+    opReturnData?: string,
+    feeRate?: number,
+    confirmationTarget: number = 3
+): Promise<bigint> {
+    const addressInfo = getAddressInfo(fromAddress);
+
+    if (addressInfo.network === 'regtest') {
+        throw new Error('Bitcoin regtest not supported');
+    }
+
+    // We need the public key to generate the redeem and witness script to spend the scripts
+    if (addressInfo.type === (AddressType.p2sh || AddressType.p2wsh)) {
+        if (!publicKey) {
+            throw new Error('Public key is required to spend from the selected address type');
+        }
+    }
+
+    // Use the from address as the toAddress for the fee estimate
+    const toAddress = fromAddress;
+
+    // TODO: allow submitting the UTXOs, fee estimate and confirmed transactions
+    // to avoid fetching them again.
+    const esploraClient = new EsploraClient(addressInfo.network);
+
+    let confirmedUtxos: UTXO[] = [];
+    if (feeRate) {
+        confirmedUtxos = await esploraClient.getAddressUtxos(fromAddress);
+    } else {
+        [confirmedUtxos, feeRate] = await Promise.all([
+            esploraClient.getAddressUtxos(fromAddress),
+            esploraClient.getFeeEstimate(confirmationTarget),
+        ]);
+    }
+
+    if (confirmedUtxos.length === 0) {
+        throw new Error('No confirmed UTXOs');
+    }
+
+    // To construct the spending transaction and estimate the fee, we need the transactions for the UTXOs
+    const possibleInputs: Input[] = [];
+
+    await Promise.all(
+        confirmedUtxos.map(async (utxo) => {
+            const hex = await esploraClient.getTransactionHex(utxo.txid);
+            const transaction = Transaction.fromRaw(Buffer.from(hex, 'hex'), { allowUnknownOutputs: true });
+            const input = getInputFromUtxoAndTx(
+                addressInfo.network as BitcoinNetworkName,
+                utxo,
+                transaction,
+                addressInfo.type,
+                publicKey
+            );
+            possibleInputs.push(input);
+        })
+    );
+
+    // Create transaction without outputs
+    const targetOutputs: Output[] = [
+        {
+            address: toAddress,
+            amount: BigInt(amount ? amount : 0),
+        },
+    ];
+
+    if (opReturnData) {
+        // Strip 0x prefix from opReturn
+        if (opReturnData.startsWith('0x')) {
```

Comment:
If this function is supposed to be used by third parties, then it might be a good idea to make this as robust as possible. Let's do this in case some maniac decides to uppercase the opreturn (making it start with `0X`)?

```suggestion
        if (opReturnData.toLowerCase().startsWith('0x')) {
```